File: internal/domain/oncall.go
package domain

import (
	"context"
	"strings"
	"time"
)

// OnCallScheduleStatus is the lifecycle of an on-call schedule. Modelled on
// TeamStatus, not on a richer state machine: a schedule is a governed,
// named object an operator archives rather than deletes so anything already
// pointing at it (a participant row, a future E5.2b escalation tier) is not
// orphaned (ADR-IDENTITY-001 §8.3's reasoning, applied here).
type OnCallScheduleStatus string

const (
	OnCallScheduleActive   OnCallScheduleStatus = "active"
	OnCallScheduleArchived OnCallScheduleStatus = "archived"
)

// Valid reports whether s is a defined status.
func (s OnCallScheduleStatus) Valid() bool {
	return s == OnCallScheduleActive || s == OnCallScheduleArchived
}

// MaxOnCallScheduleNameLength bounds a free-text field that reaches consoles
// and logs, the same bound Team.Name carries.
const MaxOnCallScheduleNameLength = 200

// MinOnCallHandoffIntervalSeconds is the floor on how often a rotation hands
// off: zero or negative would make the rotation math (OnCallRotationIndex)
// divide by a non-positive duration, which is nonsensical rather than merely
// unusual, so it is refused outright rather than clamped.
const MinOnCallHandoffIntervalSeconds = 1

// OnCallSchedule is a named on-call rotation: a set of participants
// (OnCallParticipant, ordered by Position) who hand the on-call duty to one
// another every HandoffIntervalSeconds, starting at RotationStartAt (E5.2a).
//
// This is a real, first-class operational object an operator declares — not
// one of the reduced false-nouns (docs/PLATFORM-BUILD-PLAN.md §4, Vol II
// §5.3), exactly the same reasoning MaintenanceWindow's own doc comment
// applies. What IS reduced-concept-sensitive is "who is on call right now":
// that is never a stored fact on this type or on any row anywhere — it is
// always recomputed from RotationStartAt/HandoffIntervalSeconds and the
// current ordered participant list by OnCallRotationIndex, a pure function.
// A per-shift assignment table (row per person per period) was the framework
// this story explicitly rejects: it would need a background job to keep
// generating future shifts, a query to find "the" current one, and a second
// source of truth that could drift from the rotation's own definition.
type OnCallSchedule struct {
	ScheduleID string
	TenantID   string
	Name       string
	// HandoffIntervalSeconds is how long each participant holds on-call duty
	// before the rotation advances to the next one, in whole seconds. All
	// rotation math is seconds-based UTC arithmetic — no calendars, no time
	// zones, no DST — a deliberate simplification (see OnCallRotationIndex's
	// doc comment); a schedule that needs "every Monday 09:00 local time"
	// semantics is out of this story's scope.
	HandoffIntervalSeconds int
	// RotationStartAt anchors the rotation: at exactly this instant (and
	// every HandoffIntervalSeconds after it) the on-call duty hands to the
	// next participant in Position order. Stored and compared in UTC.
	RotationStartAt time.Time
	Status          OnCallScheduleStatus
	RowVersion      int64
	CreatedAt       time.Time
	UpdatedAt       time.Time
}

// Active reports whether the schedule is in ordinary use.
func (s *OnCallSchedule) Active() bool { return s.Status == OnCallScheduleActive }

// Validate enforces the invariants the database also enforces, so a bad
// request fails with a field-level 422 rather than a constraint-violation 500.
func (s *OnCallSchedule) Validate() error {
	if strings.TrimSpace(s.ScheduleID) == "" {
		return newValidation("schedule_id", "must not be empty")
	}
	if strings.TrimSpace(s.TenantID) == "" {
		return newValidation("tenant_id", "must not be empty; it is the row's isolation key")
	}
	name := strings.TrimSpace(s.Name)
	if name == "" {
		return newValidation("name", "must not be empty")
	}
	if len(name) > MaxOnCallScheduleNameLength {
		return newValidation("name", "must be at most 200 characters")
	}
	if s.HandoffIntervalSeconds < MinOnCallHandoffIntervalSeconds {
		return newValidation("handoff_interval_seconds", "must be greater than 0")
	}
	if s.RotationStartAt.IsZero() {
		return newValidation("rotation_start_at", "must not be zero")
	}
	if !s.Status.Valid() {
		return newValidation("status", "must be one of: active, archived")
	}
	return nil
}

// NewOnCallSchedule builds a freshly minted, active schedule.
func NewOnCallSchedule(
	tenantID, name string, handoffIntervalSeconds int, rotationStartAt time.Time,
) (*OnCallSchedule, error) {
	s := &OnCallSchedule{
		ScheduleID:             NewID(),
		TenantID:               strings.TrimSpace(tenantID),
		Name:                   strings.TrimSpace(name),
		HandoffIntervalSeconds: handoffIntervalSeconds,
		RotationStartAt:        rotationStartAt.UTC(),
		Status:                 OnCallScheduleActive,
	}
	if err := s.Validate(); err != nil {
		return nil, err
	}
	return s, nil
}

// OnCallSchedulePatch changes one or more fields of a schedule under
// optimistic locking. Every field is a pointer so an absent field is
// distinguishable from an explicit zero value — the same shape
// AlertRulePatch uses.
type OnCallSchedulePatch struct {
	Name                   *string
	HandoffIntervalSeconds *int
	RotationStartAt        *time.Time
	Status                 *OnCallScheduleStatus
}

// OnCallParticipant is one person's seat in a schedule's rotation order.
// Position is 0-based and contiguous within a schedule (0..N-1): AddParticipant
// appends at N, RemoveParticipant closes the resulting gap, and
// ReorderParticipants rewrites the whole sequence atomically — see
// OnCallScheduleRepository's doc comment on each. It is a durable roster
// entry, never a per-shift/per-period row: there is exactly one row per
// (schedule, participant) regardless of how many times the rotation has
// cycled past them.
type OnCallParticipant struct {
	ParticipantID string
	TenantID      string
	ScheduleID    string
	// UserID names a person the caller's tenant actually knows — re-verified
	// against an ACTIVE membership row before this row is written
	// (OnCallScheduleRepository.AddParticipant's doc comment), the identical
	// defense IncidentStore.verifyAssigneeExists applies to Incident's
	// assignee (ADR-IDENTITY-002 §3.1: app_user itself carries no tenant_id).
	UserID     string
	Position   int
	RowVersion int64
	CreatedAt  time.Time
	UpdatedAt  time.Time
}

// Validate enforces the invariants the database also enforces.
func (p *OnCallParticipant) Validate() error {
	if strings.TrimSpace(p.ParticipantID) == "" {
		return newValidation("participant_id", "must not be empty")
	}
	if strings.TrimSpace(p.TenantID) == "" {
		return newValidation("tenant_id", "must not be empty; it is the row's isolation key")
	}
	if strings.TrimSpace(p.ScheduleID) == "" {
		return newValidation("schedule_id", "must not be empty")
	}
	if strings.TrimSpace(p.UserID) == "" {
		return newValidation("user_id", "must not be empty")
	}
	if p.Position < 0 {
		return newValidation("position", "must not be negative")
	}
	return nil
}

// NewOnCallParticipant builds a participant at the given position. The
// repository, not this constructor, decides what that position is (append at
// the current count) — see OnCallScheduleRepository.AddParticipant.
func NewOnCallParticipant(scheduleID, tenantID, userID string, position int) (*OnCallParticipant, error) {
	p := &OnCallParticipant{
		ParticipantID: NewID(),
		TenantID:      strings.TrimSpace(tenantID),
		ScheduleID:    strings.TrimSpace(scheduleID),
		UserID:        strings.TrimSpace(userID),
		Position:      position,
	}
	if err := p.Validate(); err != nil {
		return nil, err
	}
	return p, nil
}

// OnCallRotationIndex answers "which of participantCount participants (by
// their 0-based rank in Position order — NOT the stored Position value
// itself, which need not be contiguous the instant this is called) holds the
// on-call seat at `at`" for a rotation beginning at rotationStartAt and
// handing off every handoffIntervalSeconds. This is the whole of E5.2a's
// "who's on call" logic: deterministic, pure, and independent of any store —
// no per-shift row is ever consulted or produced.
//
// ok is false only when participantCount <= 0: nobody is on call, which is a
// valid, ordinary state (an empty rotation), never an error.
//
// at precedes rotationStartAt: the rotation has not started, so the first
// participant (index 0) holds the seat — the same "not yet begun" convention
// a not-yet-active resource defaults to its initial state under.
//
// Otherwise: elapsed = at - rotationStartAt; steps = floor(elapsed /
// handoffIntervalSeconds); index = steps mod participantCount, using a
// POSITIVE modulo (Go's % can return a negative result for a negative
// dividend — steps is never negative here since at >= rotationStartAt on
// this branch, but the positive-modulo form is applied unconditionally so
// the function stays correct if that invariant ever changes upstream).
//
// The interval is HALF-OPEN, [rotationStartAt + k*interval,
// rotationStartAt + (k+1)*interval): at exactly a handoff boundary the
// rotation has already advanced — steps divides evenly and the NEXT
// participant holds the seat — mirroring the half-open convention
// MaintenanceWindow's [starts_at, ends_at) and TelemetryRepository.QueryRange's
// [from, to) already use in this codebase, so a boundary instant has exactly
// one owner everywhere it appears.
//
// All arithmetic is UTC and seconds-based: no calendar, no timezone, no DST.
// A rotation that needs "hand off every Monday at 09:00 local time" is out of
// this story's scope — see docs/adr/ADR-ONCALL-001's Decision on this point.
func OnCallRotationIndex(
	participantCount int, rotationStartAt time.Time, handoffIntervalSeconds int, at time.Time,
) (index int, ok bool) {
	if participantCount <= 0 {
		return 0, false
	}
	if handoffIntervalSeconds <= 0 {
		// Defensive only: Validate refuses to persist a non-positive interval,
		// so a live schedule never reaches this branch. Guarding it here
		// avoids a division-by-zero panic (time.Duration is int64; dividing
		// by a zero Duration panics) rather than trusting every caller to
		// have validated first.
		return 0, true
	}
	if at.Before(rotationStartAt) {
		return 0, true
	}

	elapsed := at.Sub(rotationStartAt)
	interval := time.Duration(handoffIntervalSeconds) * time.Second
	steps := int64(elapsed / interval)
	n := int64(participantCount)
	return int(positiveModulo(steps, n)), true
}

// positiveModulo returns x mod n in [0, n) — unlike Go's own %, which returns
// a negative result for a negative dividend (e.g. -1 % 3 == -1, not 2).
//
// Extracted to its own function, rather than inlined in OnCallRotationIndex,
// so its correctness is directly unit-testable with a negative x: the
// `at.Before(rotationStartAt)` branch above keeps every live caller's steps
// non-negative today, which would otherwise make a mutation of this formula
// (e.g. dropping back to a bare `x % n`) pass every OnCallRotationIndex test
// unchanged — a defensive guard that nothing in the public function's own
// test suite could prove is load-bearing. Testing this helper directly with
// negative x closes that gap.
func positiveModulo(x, n int64) int64 {
	return ((x % n) + n) % n
}

// OnCallScheduleRepository administers on-call schedules, their participant
// rosters, and the "who's on call now" computation (E5.2a).
//
// on_call_schedule and on_call_participant are TENANT-OWNED: both are in
// TenantOwnedTables and carry row-level security, so this repository takes no
// tenant argument anywhere — the bound connection already is the boundary
// (ADR-TENANCY-002), exactly like TeamRepository. There is no privileged-pool
// counterpart of this type: unlike AlertRuleRepository/MaintenanceWindowRepository,
// nothing in E5.2a runs as a cross-tenant background worker — "who's on call"
// is always answered on behalf of exactly one already-known tenant, at
// request time (ADR-ONCALL-001's Decision on this point).
type OnCallScheduleRepository interface {
	Create(ctx context.Context, s *OnCallSchedule) (*OnCallSchedule, error)
	Get(ctx context.Context, scheduleID string) (*OnCallSchedule, error)
	// List returns a page of the caller's schedules, keyset-paginated over
	// schedule_id.
	List(ctx context.Context, limit int, after string) ([]*OnCallSchedule, error)
	// Update changes one or more fields under optimistic locking. Returns
	// ErrVersionMismatch on a stale rowVersion, ErrNotFound if scheduleID does
	// not name a schedule visible to the caller's tenant.
	Update(ctx context.Context, scheduleID string, rowVersion int64, patch OnCallSchedulePatch) (*OnCallSchedule, error)
	// Delete removes a schedule; ON DELETE CASCADE removes its participants
	// with it (a rotation's roster has no standalone meaning once the
	// rotation itself is gone).
	Delete(ctx context.Context, scheduleID string) error

	// AddParticipant appends userID to scheduleID's roster at the next
	// position (the current participant count — 0 for the first participant).
	// userID must already be an ACTIVE member of the caller's tenant —
	// re-verified against this store's own tenant-scoped connection before
	// the row is written, mirroring IncidentStore.verifyAssigneeExists
	// exactly — a user who is not an active member returns ErrNotFound.
	// Adding a user already on this schedule returns ErrConflict: unlike
	// TeamMembership (unordered, so a duplicate add is a harmless no-op),
	// a participant carries a position, and silently ignoring a second add
	// would leave it ambiguous whether the caller also meant to reorder.
	AddParticipant(ctx context.Context, scheduleID, userID string) (*OnCallParticipant, error)
	// RemoveParticipant removes one participant and closes the resulting gap
	// in Position (every remaining participant after the removed one shifts
	// down by one), keeping positions contiguous 0..N-1. Returns ErrNotFound
	// if participantID does not name a participant of scheduleID.
	RemoveParticipant(ctx context.Context, scheduleID, participantID string) error
	// ListParticipants returns a page of scheduleID's roster in rotation
	// order (Position ascending), keyset-paginated over Position itself
	// (after is the last Position seen, "" meaning from the start) rather
	// than over ParticipantID — the rotation's own order IS the thing being
	// paged, unlike every other list in this codebase whose keyset happens to
	// coincide with insertion order.
	ListParticipants(ctx context.Context, scheduleID string, limit int, after string) ([]*OnCallParticipant, error)
	// ReorderParticipants atomically replaces scheduleID's rotation order:
	// participantIDs must be exactly the schedule's current participant set,
	// in the desired new order — no more, no fewer, and every element already
	// a participant of this schedule. On success every participant's Position
	// is rewritten to its new 0-based rank and Position stays contiguous
	// 0..N-1 throughout. A non-matching set is refused with a validation
	// error before anything is written.
	ReorderParticipants(ctx context.Context, scheduleID string, participantIDs []string) ([]*OnCallParticipant, error)

	// OnCall answers "who is on call for scheduleID at `at`" by applying
	// OnCallRotationIndex to the schedule's own RotationStartAt/
	// HandoffIntervalSeconds and its current ordered participant list. It is
	// a read, not a stored fact: nothing is written by calling this. Returns
	// ErrNotFound if scheduleID does not name a schedule visible to the
	// caller's tenant; an empty roster is NOT an error — the returned
	// OnCallResolution has a nil UserID.
	OnCall(ctx context.Context, scheduleID string, at time.Time) (*OnCallResolution, error)
}

// OnCallResolution is the answer to "who is on call for a schedule right
// now (or at some other instant)". UserID/DisplayName are both nil when the
// schedule has no participants — an ordinary, non-error state.
type OnCallResolution struct {
	ScheduleID  string
	At          time.Time
	UserID      *string
	DisplayName *string
}

File: web/src/onCall.ts
import { deleteJSON, getJSON, patchJSON, postJSON } from './api';

// Typed contract for GET/POST/PATCH/DELETE /v1/admin/on-call-schedules{,/{id},
// /{id}/on-call,/{id}/participants{,/{participantId},/reorder}}. Mirrors
// internal/httpapi/handlers_oncall.go's onCallScheduleDTO/onCallParticipantDTO/
// onCallNowDTO exactly — field names and shape, not a paraphrase. Kept
// hand-written, like incidents.ts/noc.ts, until a second consumer of the
// generated spec appears.
//
// Contract facts confirmed against the Go before E-ACT.4 was written
// (handlers_oncall.go, domain/oncall.go):
//
// - POST /v1/admin/on-call-schedules (createOnCallScheduleRequest): name/
//   handoff_interval_seconds/rotation_start_at, all required.
//   schedule_id is minted server-side; status is always "active" at creation
//   (there is no caller-chosen initial status).
// - PATCH /v1/admin/on-call-schedules/{id} (patchOnCallScheduleRequest):
//   **requires** row_version (rejected < 1 as a 422); name/
//   handoff_interval_seconds/rotation_start_at/status are all independently
//   optional pointers — at least one must be set. Stale row_version is 409.
// - DELETE /v1/admin/on-call-schedules/{id} exists
//   (domain.OnCallScheduleRepository.Delete has no row_version parameter,
//   the same asymmetry ADR-ACT-002/ADR-ACT-004 already found for
//   deleteAlertRule/deleteMaintenanceWindow) but is NOT wired into this
//   console: OnCallSchedule's own doc comment states a schedule "is a
//   governed, named object an operator archives rather than deletes" —
//   `status: "archived"` via PATCH is the intended retirement path, not
//   DELETE. No deleteOnCallSchedule function exists here by design.
// - POST .../participants (addOnCallParticipantRequest): user_id only;
//   position is always "append at the end", never caller-chosen.
//   **No row_version anywhere in this request** — AddParticipant takes no
//   optimistic-lock parameter. 404 means user_id does not name an ACTIVE
//   member of the caller's tenant; 409 means user_id is already on this
//   roster (a business-rule conflict, not a stale-read conflict — the
//   response detail says so directly, and this console surfaces it as an
//   ordinary inline error rather than the row_version "changed since you
//   loaded it" refetch banner ADR-ACT-001 established for schedule PATCH).
// - DELETE .../participants/{participantId}: **no row_version, no body** —
//   RemoveParticipant takes no optimistic-lock parameter either.
// - POST .../participants/reorder (reorderOnCallParticipantsRequest):
//   participant_ids names the schedule's FULL current participant set, in
//   the desired new order — no more, no fewer. **No row_version** —
//   ReorderParticipants takes no optimistic-lock parameter. A mismatched set
//   is refused with a 422 before anything is written (atomic: either every
//   position moves or none do).
//
// onCallParticipantDTO does carry its own row_version field (every DTO in
// this package does), but none of Add/Remove/Reorder Participant currently
// consume it — it is read-only here today, not a fabricated round-trip.

export const ON_CALL_SCHEDULE_STATUSES = ['active', 'archived'] as const;
export type OnCallScheduleStatus = (typeof ON_CALL_SCHEDULE_STATUSES)[number];

export interface OnCallScheduleDTO {
  schedule_id: string;
  name: string;
  handoff_interval_seconds: number;
  rotation_start_at: string;
  status: OnCallScheduleStatus;
  row_version: number;
  created_at: string;
  updated_at: string;
}

export interface OnCallParticipantDTO {
  participant_id: string;
  schedule_id: string;
  user_id: string;
  position: number;
  row_version: number;
  created_at: string;
  updated_at: string;
}

/** user_id/display_name are both absent when the schedule has no participants — an ordinary, non-error state, not a 404. */
export interface OnCallNowDTO {
  schedule_id: string;
  at: string;
  user_id?: string;
  display_name?: string;
}

/** The bound on the schedule list fetch — one page, matching every other board's list cap (incidents.ts' INCIDENT_LIST_CAP, alertRules.ts' ALERT_RULE_LIST_CAP). */
export const ON_CALL_SCHEDULE_LIST_CAP = 100;

/**
 * The bound on the per-schedule supplementary fetches this board issues
 * after the schedule list loads (`/on-call` and `/participants`, two
 * requests per schedule, run in parallel via Promise.allSettled) — an N+1
 * pattern like api.ts's `getRelations`/`RESOLVE_LIMIT`, capped smaller here
 * because it is two round trips per schedule rather than one. A schedule
 * beyond this cap still renders (name, handoff interval) with an explicit
 * "not loaded" state for on-call/participants rather than the board issuing
 * an unbounded number of parallel requests.
 */
export const ON_CALL_DETAIL_FETCH_CAP = 20;

export function listOnCallSchedules(
  signal?: AbortSignal,
): Promise<{ items: OnCallScheduleDTO[] }> {
  const p = new URLSearchParams();
  p.set('limit', String(ON_CALL_SCHEDULE_LIST_CAP));
  return getJSON<{ items: OnCallScheduleDTO[] }>(`/v1/admin/on-call-schedules?${p.toString()}`, signal);
}

export function getOnCallNow(scheduleId: string, signal?: AbortSignal): Promise<OnCallNowDTO> {
  return getJSON<OnCallNowDTO>(
    `/v1/admin/on-call-schedules/${encodeURIComponent(scheduleId)}/on-call`,
    signal,
  );
}

export function listOnCallParticipants(
  scheduleId: string,
  signal?: AbortSignal,
): Promise<{ items: OnCallParticipantDTO[] }> {
  const p = new URLSearchParams();
  p.set('limit', '100');
  return getJSON<{ items: OnCallParticipantDTO[] }>(
    `/v1/admin/on-call-schedules/${encodeURIComponent(scheduleId)}/participants?${p.toString()}`,
    signal,
  );
}

export function getOnCallSchedule(scheduleId: string, signal?: AbortSignal): Promise<OnCallScheduleDTO> {
  return getJSON<OnCallScheduleDTO>(`/v1/admin/on-call-schedules/${encodeURIComponent(scheduleId)}`, signal);
}

export interface CreateOnCallScheduleInput {
  name: string;
  handoff_interval_seconds: number;
  rotation_start_at: string;
}

/** POST /v1/admin/on-call-schedules — createOnCallScheduleRequest. schedule_id is minted server-side. */
export function createOnCallSchedule(
  input: CreateOnCallScheduleInput,
  signal?: AbortSignal,
): Promise<OnCallScheduleDTO> {
  return postJSON<OnCallScheduleDTO>('/v1/admin/on-call-schedules', input, signal);
}

export interface OnCallSchedulePatchInput {
  name?: string;
  handoff_interval_seconds?: number;
  rotation_start_at?: string;
  status?: OnCallScheduleStatus;
}

/**
 * PATCH /v1/admin/on-call-schedules/{id} — patchOnCallScheduleRequest.
 * row_version is the value the last GET of this schedule returned; a stale
 * value throws an ApiError with status 409 (see this module's top-of-file
 * contract note).
 */
export function patchOnCallSchedule(
  scheduleId: string,
  patch: OnCallSchedulePatchInput,
  rowVersion: number,
  signal?: AbortSignal,
): Promise<OnCallScheduleDTO> {
  return patchJSON<OnCallScheduleDTO>(
    `/v1/admin/on-call-schedules/${encodeURIComponent(scheduleId)}`,
    { ...patch, row_version: rowVersion },
    signal,
  );
}

/**
 * POST .../participants — addOnCallParticipantRequest. Appends userId at the
 * end of the rotation. No row_version (see this module's top-of-file
 * contract note); a 404 means userId does not name an ACTIVE member of this
 * tenant, a 409 means userId is already on this roster.
 */
export function addOnCallParticipant(
  scheduleId: string,
  userId: string,
  signal?: AbortSignal,
): Promise<OnCallParticipantDTO> {
  return postJSON<OnCallParticipantDTO>(
    `/v1/admin/on-call-schedules/${encodeURIComponent(scheduleId)}/participants`,
    { user_id: userId },
    signal,
  );
}

/** DELETE .../participants/{participantId}. No row_version, no body (see this module's top-of-file contract note). */
export function removeOnCallParticipant(
  scheduleId: string,
  participantId: string,
  signal?: AbortSignal,
): Promise<void> {
  return deleteJSON(
    `/v1/admin/on-call-schedules/${encodeURIComponent(scheduleId)}/participants/${encodeURIComponent(participantId)}`,
    signal,
  );
}

/**
 * POST .../participants/reorder — reorderOnCallParticipantsRequest.
 * participantIds must be exactly the schedule's current participant set, in
 * the desired new order (this module's top-of-file contract note); the
 * caller (components/OnCallScheduleDetail.tsx) always sends the full roster
 * it just rendered, never a partial move, so the "atomic set, not N
 * interleaved moves" hard constraint holds by construction.
 */
export function reorderOnCallParticipants(
  scheduleId: string,
  participantIds: string[],
  signal?: AbortSignal,
): Promise<{ items: OnCallParticipantDTO[] }> {
  return postJSON<{ items: OnCallParticipantDTO[] }>(
    `/v1/admin/on-call-schedules/${encodeURIComponent(scheduleId)}/participants/reorder`,
    { participant_ids: participantIds },
    signal,
  );
}

/** `domain.MaxOnCallScheduleNameLength`, restated for client-side validation. */
export const MAX_ON_CALL_SCHEDULE_NAME_LENGTH = 200;

/** `undefined` when v is empty (not yet answered) or valid; an error string otherwise. */
export function onCallScheduleNameError(v: string): string | undefined {
  return v.length > MAX_ON_CALL_SCHEDULE_NAME_LENGTH
    ? `Must be at most ${MAX_ON_CALL_SCHEDULE_NAME_LENGTH} characters.`
    : undefined;
}

/** `domain.MinOnCallHandoffIntervalSeconds`, restated: a rotation must hand off at least once a second. */
export function onCallHandoffIntervalError(seconds: number): string | undefined {
  return Number.isFinite(seconds) && seconds >= 1 ? undefined : 'Must be greater than 0.';
}

/**
 * Friendly handoff-interval presets offered before falling back to a raw
 * "custom seconds" field — the brief's own suggestion (1 day/1 week), so an
 * operator declaring an ordinary rotation never has to do the seconds
 * arithmetic themselves.
 */
export const ON_CALL_HANDOFF_PRESETS: { value: string; label: string; seconds?: number }[] = [
  { value: '86400', label: '1 day', seconds: 86_400 },
  { value: '604800', label: '1 week', seconds: 604_800 },
  { value: 'custom', label: 'Custom (seconds)' },
];

/** The preset matching an existing schedule's interval, or 'custom' if it matches none of them. */
export function presetForHandoffSeconds(seconds: number): string {
  const match = ON_CALL_HANDOFF_PRESETS.find((p) => p.seconds === seconds);
  return match ? match.value : 'custom';
}
